### ISSUE TITLE
Hooking into file file writing

### ISSUE THREAD

**Issue by marijnh** [rollup/rollup#3150]
### Feature Use Case

What I'm trying to do is avoid writing a file when the file already exists, with the same content, on disk. This is mostly useful for TypeScript declaration file output produced by a typescript plugin, which usually doesn't change between recompiles, but which triggers other watching rebuilds when overwritten.

I can do this when calling `rollup.rollup`, since there I can just get the output and write it myself, but it seems `rollup.watch` requires me to leave the file writing to rollup itself, and, going by the source code, it seems to unconditionally call `fs.writeFile` there.

### Feature Proposal

I'd be happy with any of:

 - An option that holds a callback that is called instead of rollup's internal `writeFile` function when given.

 - A plugin hook that does something like that.

 - An option (or even default behavior) that does exactly what I describe above—don't write files when they already exists and have the same content.

If any of those would be welcome, I can create a pull request.

**Comment by lukastaegert** [comment]
I think what you want to achieve is already possible using the `generateBundle` hook. During that hook, you have the generated files available in memory and can compare  them with the files on disk, and then just delete files from the bundle object that you do not want to emit. See also the hint at the end of https://rollupjs.org/guide/en/#generatebundle about deleting from the bundle.

**Comment by marijnh** [comment]
Ahh, indeed, I hadn't seen that deleting files was allowed. That gets me what I needed, thanks.

**Comment by marijnh** [comment]
This is another context than the originally described case, but it seems the bundle object used by `reserveFileNameInBundle` is _not_ the bundle object passed to `generateBundle`, so deleting a property in that hook doesn't actually remove it from the bundle entirely—when you try to emit it again, you get an error (`Could not emit file "${fileName}" as it conflicts with an already emitted file.`). Is that intentional? Is there any way around it?

**Comment by lukastaegert** [comment]
I think that would be a useful feature request I did not consider yet. Would it be possible to add an update to the issue description so that it is clear why this issue is open? Alternatively, it might make sense to move this to a separate issue.

**Comment by marijnh** [comment]
Thanks for the response. Unfortunately, the issue we're having (https://github.com/ezolenko/rollup-plugin-typescript2/issues/180) isn't really covered by the 'hook into writing a file' scenario that I started this issue with (so a new issue would definitely be appropriate). But before I try to frame it as an issue proposal, I'd like a little feedback that you're probably the best person to provide. Summarizing the issue I linked:

 - rollup-plugin-typescript2 currently writes .d.ts files, which are a kind of by-product of compiling typescript files, directly to disk itself

 - It seems like a good idea to use Rollup's emit functionality for this. Concretely, this removes the issue where using the plugin through the `bundle.generate` API method breaks generating .d.ts files.

 - But doing that in the simple way, just replacing the file write call with `emitFile`, we keep getting the "conflicts with an already emitted file" error when there are multiple output targets for a bundle, and can't find a good way around it.

So the first questions would be why is `generateBundle` being called multiple times with the same 'checking for duplicate files' namespace? Is the plugin somehow supposed to know it shouldn't re-emit some files, or is this an issue in Rollup? Would adding an option to `emitFile` to just always overwrite existing files by the given name be a reasonable solution here, or would that introduce other problems?

**Comment by marijnh** [comment]
The issue only shows up when there are several outputs. The files are emitted from the `generateBundle` hook.

That hook is called only once for each output, _but_ when it emits files by the same name for each of the outputs, we get the conflict error. Since the typescript declaration files are going to be the same for each output, that is intentional (the current code will write the same file to disk multiple times, which is wasteful, but okay)—even if you emit cjs and esm code, TypeScript is only going to consume one kind of declaration file for the module.

If, instead, the plugin keeps track of what has already been emitted and only emits the declaration files once, it'll fail to re-emit the files in watch mode.

> Additionally we could also add an option to suppress this warning for a given file emission, but I am not sure this is really in everyone's best interest.

I think this is what we need here—it is definitely reasonable to error (or warn) by default in this case, since I agree it's a bit fishy, but when a plugin has some side product files like these, I think just re-emitting them for each output is the reasonable, simple thing to do, even if it's a bit redundant.

**Comment by lukastaegert** [comment]
Well, as stated, my original intention was that there should be no interaction at all between files emitted from different `generateBundle` calls, so this should definitely be looked into. As I understand, fixing that would already solve your issue, wouldn't it? Then maybe we can get away with that + replacing the errors with warnings for good measure.

**Comment by marijnh** [comment]
> As I understand, fixing that would already solve your issue, wouldn't it?

It would, but I feel that might be overkill—it is probably rare for different outputs to emit overlapping files, so allowing it through some opt-in option might be reasonable.

Would putting an option related to this in the object passed to `emitFile` be reasonable, or would you prefer to add an extra optional option-object argument to that method?

**Comment by marijnh** [comment]
> Actually the intention is that each output is a completely separate virtual directory

That is a good point. Agreed then, let's look into fixing that. Opening an issue.
